mopidy/files/library.py

```diff
@@ -0,0 +1,136 @@
+from __future__ import unicode_literals
+
+import logging
+import operator
+import os
+import stat
+import sys
+import urllib2
+
+from mopidy import backend, exceptions, models
+from mopidy.audio import scan, utils
+from mopidy.internal import path
+
+logger = logging.getLogger(__name__)
+
+
+class FilesLibraryProvider(backend.LibraryProvider):
+    """Library for browsing local files."""
+    # TODO: get_images that can pull from metadata and/or .folder.png etc?
+    # TODO: handle playlists?
+
+    @property
+    def root_directory(self):
+        if not self._media_dirs:
+            return None
+        elif len(self._media_dirs) == 1:
+            local_path = self._media_dirs[0]['path']
+            uri = path.path_to_uri(local_path)
+        else:
+            uri = u'file:root'
+        return models.Ref.directory(name='Files', uri=uri)
+
+    def __init__(self, backend, config):
+        super(FilesLibraryProvider, self).__init__(backend)
+        self._media_dirs = list(self._get_media_dirs(config))
+        self._follow_symlinks = config['files']['follow_symlinks']
+        self._show_dotfiles = config['files']['show_dotfiles']
+        self._scanner = scan.Scanner(
+            timeout=config['files']['metadata_timeout'])
+
+    def browse(self, uri):
+        logger.debug('browse called with uri %s', uri)
+        result = []
+        local_path = path.uri_to_path(uri)
+        if local_path == 'root':
+            return list(self._get_media_dirs_refs())
+        for name in os.listdir(local_path):
+            if not self._is_in_basedir(local_path):
+                logger.warn(u'Not in base_dir: %s', local_path)
+                continue
+            child = os.path.join(local_path, name)
+            logger.debug('child: %s', child)
+            uri = path.path_to_uri(child)
+            name = name.decode(sys.getfilesystemencoding(), 'ignore')
+            if not self._show_dotfiles and name.startswith(b'.'):
+                continue
+            if self._follow_symlinks:
+                st = os.stat(child)
+            else:
+                st = os.lstat(child)
+            if stat.S_ISDIR(st.st_mode):
+                result.append(models.Ref.directory(name=name, uri=uri))
+            elif stat.S_ISREG(st.st_mode) and self._is_audiofile(uri):
+                result.append(models.Ref.track(name=name, uri=uri))
+            else:
+                logger.warn('Ignored file: %s',
+                            child.decode(sys.getfilesystemencoding(),
+                                         'ignore'))
+                continue
```

Comment:
Instead of using `stat()`, what about this:

``` python
printable_path = child.decode(sys.getfilesystemencoding(), 'ignore')

if os.path.islink(child) and not self._follow_symlinks:
    logger.debug('Ignoring symlink: %s', printable_path)
    continue

if not self._is_in_basedir(os.path.realpath(child)):
    logger.debug('Ignoring symlink to outside base dir: %s', printable_path)
    continue

if os.path.isdir(child):
    result.append(models.Ref.directory(name=name, uri=uri))
elif os.path.isfile(child):
    if self._is_audiofile(uri):
        result.append(models.Ref.track(name=name, uri=uri))
    else:
        logger.debug('Ignoring non-audiofile: %s', printable_path)
```

In addition to the existing functionality, this checks that each symlink in the dir points to something inside the media_dirs.

I also lowered the severity of the last log message from warn to debug and removed the redundant `continue` on the last line.